File: backend/skills/extractor.py
import google.generativeai as genai
import os
import json
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

def setup_gemini():
    api_key = os.getenv("GEMINI_API_KEY")
    if api_key:
        genai.configure(api_key=api_key)
        return True
    return False

def extract_claims(text: str) -> list:
    """
    Extract factual claims from a block of text.
    Returns a list of string claims.
    """
    if not setup_gemini():
        logger.warning("GEMINI_API_KEY not found. Using mock extractor.")
        # Simple mock extraction by splitting sentences roughly
        return [c.strip() for c in text.split('.') if len(c.strip()) > 5]

    try:
        model = genai.GenerativeModel('gemini-1.5-flash')
        prompt = f"""
        Analyze the following text and extract all verifiable factual claims.
        Break down complex sentences into individual atomic claims.
        Return ONLY a valid JSON array of strings. Do not include markdown formatting or backticks.
        
        Text:
        {text}
        """
        response = model.generate_content(prompt)
        text_resp = response.text.strip()
        if text_resp.startswith('```json'):
            text_resp = text_resp[7:-3]
        elif text_resp.startswith('```'):
            text_resp = text_resp[3:-3]
            
        claims = json.loads(text_resp)
        return claims
    except Exception as e:
        logger.error(f"Error in extract_claims: {e}")
        return [text]

File: node-backend/skills/extractor.js
const Groq = require("groq-sdk");

async function extractClaims(text) {
  const prompt = `
    Analyze the following text and extract all verifiable factual claims.
    CRITICAL INSTRUCTION: Do NOT overly fragment sentences into isolated, obvious facts (e.g., do not split "India is located in Australia" into "India is a country" and "Australia is a continent"). Instead, extract the main relational claim: "India is located in Australia". Focus on the core relationships, statistics, geographical statements, or events being asserted.
    Return ONLY a valid JSON array of strings. Do not include markdown formatting or backticks.
    
    Text:
    ${text}
  `;

  try {
    if (!process.env.GROQ_API_KEY) {
      throw new Error("GROQ_API_KEY is not set.");
    }
    const groq = new Groq({ apiKey: process.env.GROQ_API_KEY });

    const chatCompletion = await groq.chat.completions.create({
      messages: [{ role: "user", content: prompt }],
      model: "llama-3.3-70b-versatile",
      temperature: 0.1,
    });

    let textResp = chatCompletion.choices[0]?.message?.content?.trim() || "";

    if (textResp.startsWith('```json')) {
      textResp = textResp.slice(7, -3);
    } else if (textResp.startsWith('```')) {
      textResp = textResp.slice(3, -3);
    }

    return JSON.parse(textResp);
  } catch (error) {
    console.error("Error extracting claims:", error);
    // Simple fallback: split by sentences
    return text.split(/[.!?]+/).filter(s => s.trim().length > 10).map(s => s.trim());
  }
}

module.exports = { extractClaims };
